src/editor/TextSearch.java

```diff
@@ -0,0 +1,57 @@
+package editor;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Map;
+import java.util.TreeMap;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+public class TextSearch {
+
+    static  ArrayList<Integer> startIndexes = new ArrayList<>();
```

Comment:
not private static modifiable fields is not a good approach in multithreaded programming because several parts of your program potentially may change the same data. You should design your classes in a way that only those parts of the program can access it who actually need it.

It  may be better to use an instance of searcher:
TextSearch searcher = new TextSearch(...). 
The instance should have its entire arrays.